include/dali/core/os/shared_mem.h

```diff
@@ -0,0 +1,158 @@
+      DALI_FAIL(make_string("Call to ", call_str, " failed. ", errmsg, " ", message)); \
+    }                                                                                  \
+  } while (0)
+
+
+#define POSIX_CHECK_STATUS(status, call_str) POSIX_CHECK_STATUS_EX(status, call_str, "")
+
+#define POSIX_CALL_EX(code, message)               \
+  do {                                             \
+    int status = code;                             \
+    POSIX_CHECK_STATUS_EX(status, #code, message); \
+  } while (0)
+
+
+#define POSIX_CALL(code) POSIX_CALL_EX(code, "")
+
+#else
+#error Platform not supported
+#endif
+
+class DLL_PUBLIC FileHandle : public UniqueHandle<fd_handle_t, FileHandle> {
+ public:
+  DALI_INHERIT_UNIQUE_HANDLE(fd_handle_t, FileHandle);
+
+  static void DestroyHandle(fd_handle_t h);
+
+  static constexpr fd_handle_t null_handle() {
+    return -1;
+  }
+};
+
+
+/**
+ * @brief Creates or simply stores (if provided) handle used for identifying
+ * shared memory chunk. Closes handle in the desctructor if it
+ * wasn't closed earlier.
+ */
+class DLL_PUBLIC ShmHandle : public UniqueHandle<shm_handle_t, ShmHandle> {
+ public:
+  DALI_INHERIT_UNIQUE_HANDLE(shm_handle_t, ShmHandle);
+
+  /**
+   * Create new handle (file descriptor on Unix) that can be used for shared memory chunk.
+   */
+  static ShmHandle CreateHandle();
+
+  static void DestroyHandle(shm_handle_t h);
+
+  static constexpr shm_handle_t null_handle() {
+    return -1;
+  }
+};
+
+
+struct MappedMemoryChunk {
+  uint64_t size;
+  uint8_t *ptr;
+  bool operator==(const MappedMemoryChunk &other) const {
+    return size == other.size && ptr == other.ptr;
+  }
+};
+
+/**
+ * @brief Map/unmap and resize mapping of shared memory into process address space.
+ * Unmaps memory in the destructor if it wasn't unmapped earlier.
+ */
+class DLL_PUBLIC MemoryMapping : public UniqueHandle<MappedMemoryChunk, MemoryMapping> {
+ public:
+  DALI_INHERIT_UNIQUE_HANDLE(MappedMemoryChunk, MemoryMapping);
+
+  MemoryMapping(shm_handle_t handle, uint64_t size);
+
+  void resize(uint64_t size);
+
+  uint8_t *get_raw_ptr();
+
+  static void DestroyHandle(MappedMemoryChunk);
+
+  static constexpr MappedMemoryChunk null_handle() {
+    return {0, nullptr};
+  }
+};
+
+
+/**
+ * @brief Allocate or access existing shared memory chunk identified by handle (file descriptor).
+ * When created with handle = -1, it allocates new shared memory of size ``size`` and mmaps
+ * it into process address space. When handle of existing shared memory chunk is
+ * passed the related shared memory chunk will be mmaped into process address space,
+ * ``size`` should match the size of underlying shared memory chunk.
+ */
+class DLL_PUBLIC SharedMem {
+ public:
+  using b_type = uint8_t;
+
+  DLL_PUBLIC SharedMem(shm_handle_t handle, uint64_t size);
+
+  DLL_PUBLIC ~SharedMem() = default;
+
+  DLL_PUBLIC uint64_t size() const;
+
+  DLL_PUBLIC shm_handle_t handle();
+
+  DLL_PUBLIC b_type *get_raw_ptr();
```

Comment:
These should be inline. There's really no point in going through DLL to return a field value.